### DISCUSSION TITLE
[WASM] WASM throws an exception, when a Azure Function is called

### DISCUSSION THREAD

**Discussion by Grzesik** [unoplatform/uno#10113] [Q&A]
Hi, I have a problem, when I call a Azure Function from WASM. It works from the Windows platform.
I created an Azure Function and started it locally:

```
[FunctionName("Function1")]
        public static async Task<IActionResult> Run(
            [HttpTrigger(AuthorizationLevel.Function, "get", "post", Route = null)] HttpRequest req,
            ILogger log)
        {
            log.LogInformation("C# HTTP trigger function processed a request.");

            string name = req.Query["name"];

            string requestBody = await new StreamReader(req.Body).ReadToEndAsync();
            dynamic data = JsonConvert.DeserializeObject(requestBody);
            name = name ?? data?.name;

            string responseMessage = string.IsNullOrEmpty(name)
                ? "This HTTP triggered function executed successfully. Pass a name in the query string or in the request body for a personalized response."
                : $"Hello, {name}. This HTTP triggered function executed successfully.";

            return new OkObjectResult(responseMessage);
        }
```

I called it from the Windows Platform and it works fine:

```
var uri = new System.Uri("http://localhost:7078/api/Function1");
            using (var httpClient = new HttpClient())
            {
                // Always catch network exceptions for async methods
                try
                {
                    var httpRequestMessage = new HttpRequestMessage(HttpMethod.Get, uri);
                    using (var response = await httpClient.SendAsync(httpRequestMessage))
                    {
                        if (response.IsSuccessStatusCode)
                        {
                            var str =  await response.Content.ReadAsStringAsync();
                        }
                    }
                    //string result = await httpClient.GetStringAsync(uri);

                }
                catch (System.Exception ex)
                {
                    // Details in ex.Message and ex.HResult.
                }
            }
```

When I call the same code from WASM I get an Exception:

_innerException = System.Runtime.InteropServices.JavaScript.JSException
_message = 'TypeError: Failed to fetch'

The Function is called and returns OK (I see it in debugger), but WASM has a problem with the result. In the Command Window, there is a folowing Message:

```
Now listening on: http://localhost:5000
Application started. Press Ctrl+C to shut down.
Hosting environment: Production
Content root path: D:\Development\DevGit\WebStaticApp1\WebStaticApp1\WebStaticApp1.Wasm
Now listening on: http://127.0.0.1:20655
Application started. Press Ctrl+C to shut down.
fail: Microsoft.WebAssembly.Diagnostics.DevToolsProxy[0] sending error response for id: msg-E3E81D1881F64833FF6D9DE4D6BB214A:::1035 -> [Result: IsOk: False, IsErr: True, Value: , Error: {   "result": {     "type": "object",     "subtype": "error",     "description": "Cannot find member named 'httpClient'.",     "className": "ReferenceError"   } } ]
fail: Microsoft.WebAssembly.Diagnostics.DevToolsProxy[0] sending error response for id: msg-E3E81D1881F64833FF6D9DE4D6BB214A:::1036 -> [Result: IsOk: False, IsErr: True, Value: , Error: {   "result": {     "type": "object",     "subtype": "error",     "description": "Cannot find member named 'httpRequestMessage'.",     "className": "ReferenceError"   } } ]
fail: Microsoft.WebAssembly.Diagnostics.DevToolsProxy[0] sending error response for id: msg-E3E81D1881F64833FF6D9DE4D6BB214A:::1037 -> [Result: IsOk: False, IsErr: True, Value: , Error: {   "result": {     "type": "object",     "subtype": "error",     "description": "Cannot find member named 'HttpMethod'.",     "className": "ReferenceError"   } } ]
fail: Microsoft.WebAssembly.Diagnostics.DevToolsProxy[0] sending error response for id: msg-E3E81D1881F64833FF6D9DE4D6BB214A:::1038 -> [Result: IsOk: False, IsErr: True, Value: , Error: {   "result": {     "type": "object",     "subtype": "error",     "description": "Failed to resolve member access for HttpMethod.Get",     "className": "ReferenceError"   } } ]
fail: Microsoft.WebAssembly.Diagnostics.DevToolsProxy[0] sending error response for id: msg-E3E81D1881F64833FF6D9DE4D6BB214A:::1039 -> [Result: IsOk: False, IsErr: True, Value: , Error: {   "result": {     "type": "object",     "subtype": "error",     "description": "Cannot find member named 'uri'.",     "className": "ReferenceError"   } } ]
fail: Microsoft.WebAssembly.Diagnostics.DevToolsProxy[0] sending error response for id: msg-E3E81D1881F64833FF6D9DE4D6BB214A:::1040 -> [Result: IsOk: False, IsErr: True, Value: , Error: {   "result": {     "type": "object",     "subtype": "error",     "description": "Cannot find member named 'this'.",     "className": "ReferenceError"   } } ]
fail: Microsoft.WebAssembly.Diagnostics.DevToolsProxy[0] sending error response for id: msg-E3E81D1881F64833FF6D9DE4D6BB214A:::1046 -> [Result: IsOk: False, IsErr: True, Value: , Error: {   "result": {     "type": "object",     "subtype": "error",     "description": "Failed to resolve member access for response.Content.ReadAsStringAsync",     "className": "ReferenceError"   } } ]

```

Is there any additional thing, I need to setup, to call the function from WASM?

**Comment by ToddGlodek** [comment]
The first thing I notice is that when calling an Azure function, you have to include the Azure access code in the query portion of the URL.    The following documentation will help you out http://dontcodetired.com/blog/post/Azure-HTTP-Function-Authorization-with-Function-Keys

Also, don't instantiate/dispose of HttpClient for every single call.  According to [Microsoft's documentation](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.httpclient?view=net-6.0) "HttpClient is intended to be instantiated once per application, rather than per-use"

If that doesn't work, consider using a GetAsync() call instead of SendAsync() - I remember reading somewhere that Send uses sockets, which are not compatible with WASM.   

Let me know how that works for you

**Reply by Grzesik** [reply]
Hi, thanks for your tip. 
1. The Key in Azure Functions is necessary, when I deploy it to Azure. It works without, when I start it locally. I tried it with Postman.
2. OK, but it is not important for the example
3. It doesn't work with GetAsync() like with SendAsync().

I guess, that WASM doesn't allow to call a local API. There have to be some settings, which allow it.

**Reply by ToddGlodek** [reply]
Yeah,   look into overcoming CORS on localhost.   You will also need to deal with the HTTP/HTTPS issues when you start writing actual code.  So you are gong to have self signed certificates to deal with.  What you want is a way to use HTTP locally for development, but then use HTTPS to connect to the actual function.

**Reply by ToddGlodek** [reply]
You might find it faster to develop using a rich client (eg WinUI) when targeting your locally hosted Azure Functions.   Your WinUI client will load so much faster than your WASM client.   Then when you are happy with the way WinUI behaves, publish the Azure Function - and contect WASM to the published version.  

I set up a compiler directive DEBUG_AZURELOCAL in the Configuration Manager.    I use the DEBUG_AZURELOCAL when debugging WinUI and  regular DEBUG when using WASM

```
#if DEBUG_AZURELOCAL
        uriBuilder = new UriBuilder("http", "localhost", 7071, $"/api/{urlPath}");        
#else

        uriBuilder = new UriBuilder("https", "myfunction123456789.azurewebsites.net") {
            Path = $"/api/{urlPath}",
            Query = ACCES_CODE
        };

#endif
```

**Reply by Grzesik** [reply]
Hi, I've published the Azure Function and tested it with Postman. I used your code, to access the function, but I get still the same exception. I is strange, because it is, like in Postmann a get with a url (with code). It is not a rocket science. I think I need a working example with uno platform / wasm, which works,  to see some settings in the project file or in some settings files.

**Reply by jeromelaban** [reply]
This kind of exception can also [be related to CORS.](https://platform.uno/docs/articles/get-started-wizard.html?q=CORS#webassembly-access-to-fetch-at-httpsxxxx-from-origin-httpxxxx-has-been-blocked-by-cors-policy). You may want to take a look at the network tab of your browser to see more details.

**Reply by Grzesik** [reply]
Yes you re right. It is a problem with CORS. I've created a Blazor-Wasm app and used the same function to call my azure function.
I added this code to the header to switch off the cors:

`httpRequestMessage.SetBrowserRequestMode(BrowserRequestMode.NoCors);`

In Blazor I get an answer, without exception, but the status is even though wrong (0). The Web Browser Console shows 200. Strange.

In UnoPlatform/Wasm  this call `httpRequestMessage.SetBrowserRequestMode(BrowserRequestMode.NoCors);` doesn't work, because it is not implemented, or I need another nuget package. Do you know, what I can do, to switch off the cors in uno platform?

**Reply by jeromelaban** [reply]
@Grzesik this is unfortunately a Blazor-only feature that is likely going to be brough down to .NET SDK, but you can probably do the same as [what is in that file](https://github.com/dotnet/aspnetcore/blob/0ee742c53f2669fd7233df6da89db5e8ab944585/src/Components/WebAssembly/WebAssembly/src/Http/WebAssemblyHttpRequestMessageExtensions.cs). If not, we may end up duplicating these extensions on Uno's side (https://github.com/unoplatform/uno/issues/10123).

Still, it's best to avoid relying on a no-cors behavior, as a lot of response information is missing. It's probably best to have your endpoint being CORS-aware.

**Reply by Grzesik** [reply]
@jeromelaban you're absolutely right, that to call an Azure-Function from WASM, it is necessary to switch off the CORS. I did it and now it works fine!

To switch off CORS locally in the Azure-Function, it is necessary to add this code to the local.settings.json:

`"Host": { "CORS": "*" }`

To switch off COSR in Azure, just go to the CORS settings in the Azure-Function and set  *  as the only option.

Thank you for help!
